docs/concepts/upgrades.md

```diff
@@ -0,0 +1,176 @@
+| | Ordered | Ordered environments require that a single offset delineates pre- and post-migration (no events are interleaved) |
+| | Live | Live environments describe queries that cannot afford downtime, either by means of acting as live storage (e.g. responding to pull queries) or feeding into high availability systems (powering important functionality) |
+
+ksqlDB supports only in-place upgrades for _data selection_ and
+_schema evolution_ upgrades on a limited subset of query characteristics.
+ksqlDB doesn't guarantee validity of any environments when performing an
+in-place upgrade.
+
+Any in place upgrades on windowed or joined sources, as well as upgrades on
+any table aggregation, are not yet supported.
+
+## A motivating example
+
+Imagine a query that reads from a stream of purchases made at ksqlDB's flagship
+store, ksqlMart, and filters out transactions that might be invalid:
+
+```sql
+CREATE STREAM purchases (product_id INT KEY, name VARCHAR, cost DOUBLE, quantity INT);
+CREATE STREAM valid_purchases AS SELECT * FROM purchases WHERE cost > 0.00 AND quantity > 0;
+```
+
+### Data selection (simple)
+
+Over time, ksqlMart changes its return policy and begins issuing full refunds.
+These events have a negative `cost` column value. Since these events are now
+valid, ksqlMart needs to update the query to remove the `cost > 0.00` clause:
+
+```sql
+CREATE OR REPLACE STREAM valid_purchases AS SELECT * FROM purchases WHERE quantity > 0;
+```
+
+This `CREATE OR REPLACE` statement instructs ksqlDB to terminate the old query,
+and create a new one with the new semantics that will continue from the last
+event that the previous query processed. Note that this means any previously 
+processed data with negative cost will not be included, even if issuing the
+query with `SET 'auto.offset.reset'='earliest';`.
+
+This query upgrade is a _simple_, _data selecting_ upgrade because it doesn't
+involve any aggregations; the only change is the criteria to emit rows. ksqlDB
+supports nearly all _data selection_ modifications on source queries.
+
+### Schema evolution
+
+Over time, ksqlMart gets more sophisticated in their usage of {{ site.ak }} to
+monitor their input. They start publishing a new field to the `purchases` stream,
+named `popularity`. In order to reflect this change in their `valid_purchases`
+stream, they need to issue two different commands:
+
+```sql
+CREATE OR REPLACE STREAM purchases (product_id INT KEY, name VARCHAR, cost DOUBLE, quantity INT, popularity DOUBLE); 
+CREATE OR REPLACE STREAM valid_purchases AS SELECT * FROM purchases WHERE quantity > 0;
+```
+
+There are a few things to note in the above statements:
+
+1. DDL statements can be updated using `CREATE OR REPLACE`.
+2. ksqlMart re-issued the `SELECT *` statement even though the statement text is
+   identical to the previous statement they issued.
+3. Why is (2) necessary? ksqlDB resolves `SELECT *` at the time the query was
```

Comment:
Bit late to the party here but wanted to say I really like this example! Great job @agavra (and @JimGalasyn ) -- this is an awesome docs topic 😄